src/main/java/org/jetbrains/bsp/bazel/server/BazelBspServer.java

```diff
@@ -187,16 +184,248 @@ public void onBuildExit() {
         });
   }
 
+  @Override
+  public CompletableFuture<SourcesResult> buildTargetSources(SourcesParams sourcesParams) {
+    return executeCommand(
+        () -> {
+          List<String> targets =
+              sourcesParams.getTargets().stream()
+                  .map(BuildTargetIdentifier::getUri)
+                  .collect(Collectors.toList());
+
+          ProcessResults processResults =
+              bazelRunner
+                  .commandBuilder()
+                  .query()
+                  .withFlag(BazelRunnerFlag.OUTPUT_PROTO)
+                  .withTargets(targets)
+                  .runBazel();
+          Build.QueryResult queryResult = QueryResolver.getQueryResultForProcess(processResults);
+
+          List<SourcesItem> sources =
+              queryResult.getTargetList().stream()
+                  .map(Build.Target::getRule)
+                  .map(
+                      rule -> {
+                        BuildTargetIdentifier label = new BuildTargetIdentifier(rule.getName());
```

Comment:
this can be extracted to a method I guess?

Reply:
I believe that isn't strictly related to this PR - there are multiple places where changes are required after this PR. I will do this in the separate PR.